### ISSUE TITLE
"http-request tarpit" appears to be broken in 1.8.x

### ISSUE THREAD

**Issue by brenc** [haproxy/haproxy#783]
# Output of `haproxy -vv` and `uname -a`

```
# haproxy -vv
HA-Proxy version 1.8.19-1+deb10u2 2020/04/01
Copyright 2000-2019 Willy Tarreau <willy@haproxy.org>

Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-ap5Gic/haproxy-1.8.19=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIF
Y_SOURCE=2 -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-forma
t-truncation -Wno-null-dereference -Wno-unused-label
  OPTIONS = USE_GETADDRINFO=1 USE_ZLIB=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 U
SE_SYSTEMD=1 USE_PCRE2=1 USE_PCRE2_JIT=1 USE_NS=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
Running on OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_F
REEBIND
Encrypted password support via crypt(3): yes
Built with multi-threading support.
Built with PCRE2 version : 10.32 2018-09-10
PCRE2 library supports JIT : yes
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw
-deflate("deflate"), gzip("gzip")
Built with network namespace support.

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available filters :
        [SPOE] spoe
        [COMP] compression
        [TRACE] trace

# uname -a
Linux staging-proxy02 4.19.0-9-amd64 #1 SMP Debian 4.19.118-2+deb10u1 (2020-06-07) x86_64 GNU/Linux
```

## What's the configuration?
```
defaults
  timeout tarpit 5s

frontend fe-main
  acl abuser url_reg -i -f /etc/haproxy/vuln_scan_urls.lst
  use_backend be-abuse if abuser

backend be-abuse
  errorfile 500 /etc/haproxy/errorfiles/500abuser.http
  http-request tarpit

# /etc/haproxy/vuln_scan_urls.lst
/\.git

```

## Steps to reproduce the behavior

Browse to `[url]/.git`.

## Actual behavior

Immediately returns a 500 error (first and all subsequent requests). Log:

```
2020-07-28T20:19:02.568942-04:00 staging-proxy02 haproxy[29018]: xx.xx.xx.xx:57023 [28/Jul/2020:20:19:02.567] fe-main~ be-abuse/<NOSRV> -1/37/-1/-1/1 500 180 - - PT-- 1/1/0/0/3 0/0 {} "GET /.git HTTP/1.1"
```

Note the `Tw` field is only 37ms.

## Expected behavior

Should pause for five seconds then return a 500. Log from 2.2.1 and 2.0.16:

```
2020-07-28T20:22:43.192123-04:00 staging-proxy02 haproxy[29326]: xx.xx.xx.xx:57927 [28/Jul/2020:20:22:38.190] fe-main~ be-abuse/<NOSRV> 0/5002/-1/-1/5002 500 188 - - PT-- 1/1/0/0/3 0/0 {} "GET /.git HTTP/2.0"
```

Note that `Tw` is reporting the expected value.

## Do you have any idea what may have caused this?

This is the first I noticed this in 1.8.x. I don't know how long this has been broken.

I compiled and tested with 1.8.25, 2.0.16, and 2.2.1 using the same config. This issue still exists in 1.8.25 but not in the others.

**Comment by wtarreau** [comment]
I can't make it fail here, for me it properly waits for 5 seconds before responding.

However your config didn't work since it's missing "mode http" (though your log shows you have it). So it may be related to anything else that's in your config and that doesn't appear in the example above. Anything else you could share that reliably reproduces it ?

**Comment by jmagnin** [comment]
This one rings a bell. I recall tarpit not working with h2 on 1.8. 
https://discourse.haproxy.org/t/timeout-tarpit-ignored-http-2/4400

**Comment by jmagnin** [comment]
here's a simple reproducer:

```
defaults
  timeout tarpit 4s
  mode http

listen in
  bind *:8443 ssl crt ssl.pem alpn h2,http/1.1
  http-request tarpit if { path /t }
  http-request deny deny_status 200
```

```
curl --http1.1 -k https://localhost:8443/t  0.01s user 0.00s system 0% cpu 4.014 total
curl --http2 -k https://localhost:8443/t  0.01s user 0.00s system 84% cpu 0.011 total
```

**Comment by wtarreau** [comment]
Ah, thanks, I didn't notice that the failing case was over H2! It could very well be caused by the limitations of the initial H2 implementation in 1.8 (H2 requests were re-encoded to H1 and fed upstream, and analysers could not really pause while processing requests). I'll have a look to confirm, or infirm and fix.

**Comment by wtarreau** [comment]
OK I found what's causing this. Sadly it's a real architectural limitation that was since addressed in 1.9. When an H2 request is encoded to H1, we have no other choice than marking the channel request closed after the request is transferred. The impact is that there was absolutely no way for the stream to know that the client hasn't aborted (this is the same reason why we couldn't support server-side keep-alive with H2 requests in 1.8). And actually, if you upload some contents larger than a buffer (more than 16kB of data), the tarpit works well because the request channel is not closed since the request is not complete.

In 1.9 and above, we've implemented a new way to mention that we've reached an end of input (CF_EOI) that is distinct from the definitive shutdown (CF_SHUTR).

I'm afraid for 1.8 there's really nothing we can do about it as it's not a bug but a technical limitation. The best we could do is to mention in the doc that it will not work with H2 requests that entirely fit in a buffer :-/
